### ISSUE TITLE
Add browser test support for logged-in applicant experience

### ISSUE THREAD

**Issue by clouser2** [civiform/civiform#3326]
Currently, the logged-in applicant experience is ONLY used when the `TEST_USER_LOGIN` / `TEST_USER_PASSWORD` environment variables are set in the test runner ([code](https://github.com/civiform/civiform/blob/ea9541238973e92f3a6c3723265be3c887d5e9f7/browser-test/src/support/index.ts#L144-L161)). Otherwise, we silently fall back to logging in as the guest.

This means that any test that make use of `loginAsTestUser` may actually be logging in as guest and not actually exercising the code paths we intend to.

Presently, we have four environments where we run tests:

1. **CI browser test runs** - These don't set the environment variable and hence don't exercise the "logged-in" flow ([code](https://github.com/civiform/civiform/blob/ea9541238973e92f3a6c3723265be3c887d5e9f7/bin/run-browser-tests-ci)).
2. **Seattle staging prober run** - The environment variables are set and passed via GitHub secrets ([code](https://github.com/civiform/civiform/blob/ea9541238973e92f3a6c3723265be3c887d5e9f7/bin/run-prober)). Additionally, our current "login" code is purpose-built for the Seattle staging auth provider ([code](https://github.com/civiform/civiform/blob/ea9541238973e92f3a6c3723265be3c887d5e9f7/browser-test/src/support/index.ts#L146-L156)).
3. **AWS staging prober run** - These don't set the environment variable and don't exercise the "logged-in" flow ([code](https://github.com/civiform/civiform-staging-deploy/blob/34f4f20e9bbb37cd577ee6bf3a19523b9efb6a1a/bin/run-prober)). Note that this invocation uses the "latest" browser-test image and the script is in a separate repo.
4. **Local browser test runs** - The environment variables aren't set in either of our local run-browser-tests script and hence don't test the logged-in experience ([code](https://github.com/civiform/civiform/blob/ea9541238973e92f3a6c3723265be3c887d5e9f7/bin/run-browser-tests)).

For status tracking, we presently will only send notification emails to the applicant if they've logged in ([code](https://github.com/civiform/civiform/blob/ea9541238973e92f3a6c3723265be3c887d5e9f7/browser-test/src/admin_application_statuses.test.ts#L165-L166)). As it stands, due to the above, there presently is no way to test this case since we don't currently support logged-in applicants for browser tests (except in Seattle staging prober runs).

This bug tracks the sequence of steps to get us to a state where we can simulate logged in flows in each of the above environments. Additionally, we'd ideally switch `loginAsTestUser` to no longer silently fall back on the guest experience and instead complain loudly if there wasn't support for logging in as a test user from whatever environment it was running in. Silently falling back isn't helping bolster test coverage.

Getting the initial bug in place and I'll hopefully describe what's needed to support this along with an implementation sequence to get us there in a subsequent update.

**Comment by clouser2** [comment]
In order to support each of the environments, we'll need to make several changes both to our application/test code as well as some of the env variables used in various environments when running browser tests. Additionally, since these changes span multiple repositories and have different deployment cadences, we'll need to incrementally enable this over the course of several PRs rather than a single PR.

## Required changes

A prototype of all of the below described changes exists at https://github.com/civiform/civiform/pull/3315/files. Note that this can't be submitted as one PR and will be split up according to the "Rollout sequencing" section below.

### browser-test/* source code

For all environments, we'll need to update the `loginAsTestUser` and `userDisplayName` methods to behave differently in each environment. The `TEST_USER_AUTH_STRATEGY` variable will be set and `loginAsTestUser` will have different adapters for Seattle staging, fake OIDC (CI / local browser test runs), and Auth0 (used in AWS staging). Additionally, `userDisplayName` will now read the "display name" for the test user from the `TEST_USER_DISPLAY_NAME` environment variable rather than hardcoding the data from Seattle's staging instance ([code](https://github.com/civiform/civiform/blob/be55897456b203aceae6dbb7df4a3e086b62e77a/browser-test/src/support/index.ts#L169)).

### Fake OIDC provider

When running browser tests, the CiviForm server is hosted at `http://civiform:9000` rather than `localhost:9000`. In browser test environments, we start an OIDC server used for login ([test_oidc_provider.js](https://github.com/civiform/civiform/blob/be55897456b203aceae6dbb7df4a3e086b62e77a/test-support/test_oidc_provider.js)) that makes use of a Node OIDC implementation (https://www.npmjs.com/package/oidc-provider).

We need to add http://civiform:9000 as a registered callback URL (https://github.com/civiform/civiform/blob/be55897456b203aceae6dbb7df4a3e086b62e77a/test-support/test_oidc_provider.js#L13-L18). This won't work out of the box since the OIDC implementation prefers not to allow redirects to insecure (i.e. `http://`) domains. Since this provider is only intended to be used for browser tests, we can follow the recipe from the [package's docs](https://github.com/panva/node-oidc-provider/blob/0fcc112e0a95b3b2dae4eba6da812253277567c9/recipes/implicit_http_localhost.md) to disable this behavior.

Additionally, we have a workaround in place for the OIDC provider not supporting "token" auth whose check needs to be extended to include redirect URLs of civiform:9000 ([code](https://github.com/civiform/civiform/blob/be55897456b203aceae6dbb7df4a3e086b62e77a/server/app/auth/oidc/applicant/IdcsProvider.java#L71-L77)).

### A test account is needed for AWS staging

We need an account that can receive email for our AWS staging probers (similar to what already exists for Seattle staging). An entry for this account can be added to our Auth0 configuration and the password can be set as a secret in the https://github.com/civiform/civiform-staging-deploy/ repo.

## Rollout sequencing

**Stage 1:** Enable for CI tests, local environment, and Seattle staging

Changes to the OIDC provider, browser test code, and associated browser test scripts can all be made in a single PR.

CI tests: Will use the fake OIDC provider, which is built afresh for every CI run. `bin/run-browser-tests-ci` will be adjusted to pass TEST_USER_AUTH_STRATEGY=fake-oidc and pass TEST_USER_ variables hardcoded to well-known values.

Local environment: Will have the same config as CI tests and the `bin/run-browser-tests` script will have the same environment variables.

Seattle staging: `bin/run-prober` will be adjusted to pass the TEST_USER_AUTH_STRATEGY=seattle-staging env variable. Additionally TEST_USER_DISPLAY_NAME will be set to "TEST, UATAPP".

AWS staging: These probers run using the "latest" tag for the browser-test image ([code](https://github.com/civiform/civiform-staging-deploy/blob/34f4f20e9bbb37cd577ee6bf3a19523b9efb6a1a/bin/run-prober#L11)). As such, we'll have to modify the associated environment variables separately and will need to wait until an image containing the 'aws-staging' strategy has been built. During the first PR submission, we'll fall back on the `loginAsGuest` behavior if the `TEST_USER_AUTH_STRATEGY` env variable is not present.

**Stage 2:** Set test user environment variables for AWS staging

Once an updated browser-test image has been built. We can modify the AWS staging prober script to set the associated environment variables.

**Stage 3:** Cleanup

Once stage 2 is complete, we can update the browser-test code to throw an error if the `TEST_USER_AUTH_STRATEGY` env variable isn't present.

**Comment by nbeloglazov** [comment]
For Stage 1. Local environment. You also need to update `bin/run-browser-tests-local` to pass the variables.

**Comment by clouser2** [comment]
Turns out that Seattle staging doesn't actually use the bin/run-prober script in the civiform/civiform repo. #3340 is out to remove that. This means that we'll need to account for Seattle staging similarly to how we account for AWS staging until a browser-test image has been built with the appropriate changes. I'm going to edit my previous comment's rollout sequence to reflect this.

**Comment by clouser2** [comment]
#3334 ended up breaking the unit test environment since we once again weren't using a locally built version of our fake OIDC provider when running unit tests (https://github.com/civiform/civiform/runs/8237822228?check_suite_focus=true). A simple test assertion needs to be updated from expecting the name to be "first" to instead be "user@example.com".

Once CI built an updated OIDC image and tagged it latest, our GitHub actions started picking this up when running unit tests. At that point, the test code had a mismatched assumption with how the provider would behave.

**Comment by clouser2** [comment]
So far it appears that the previous PR hasn't caused issue with prober runs and we have a green CI dashboard.

I'm going to proceed to Stage 2 and start with the AWS staging environment. Once that's stable, I'll coordinate to get the Seattle staging environment working.

**Comment by clouser2** [comment]
I manually tested AWS staging probers by doing the following:
  * Created a mapping for our test user email in our Auth0 configuration (username / password will eventually be added to GitHub secrets)
  * Invoked the prober script that our [staging repo uses](https://github.com/civiform/civiform-staging-deploy/blob/main/bin/run-prober)
    ```
    docker run -e TEST_USER_AUTH_STRATEGY="aws-staging" -e TEST_USER_LOGIN="<elided>" -e TEST_USER_PASSWORD='<elided>' -e TEST_USER_DISPLAY_NAME="<elided>" -e BASE_URL="https://staging-aws.civiform.dev" -e DISABLE_SCREENSHOTS=true civiform/civiform-browser-test:latest -c "yarn install && yarn probe"
    ```
  * Observed a successful run:
    ```
    ...other test output...
    PASS browser: chromium src/landing_page.test.ts (5.324 s)
    
    Test Suites: 37 passed, 37 total
    Tests:       1 skipped, 227 passed, 228 total
    Snapshots:   0 total
    Time:        1879.707 s
    ```

Given this, the next steps are to:
  1) Add the TEST_USER_LOGIN / TEST_USER_PASSWORD to the GitHub secrets for the https://github.com/civiform/civiform-staging-deploy/ repo
  2) Modify the civiform-staging-deploy repo's [`bin/run-prober`](https://github.com/civiform/civiform-staging-deploy/blob/main/bin/run-prober) script to set the appropriate environment variables
  3) Observe a successful prober run

Once that's done and the prober has been stable for a few days, we can clean up by removing the fallback code in our browser test helpers (already tracked by a TODO).

**Comment by clouser2** [comment]
Manually kicked off an AWS staging prober run, which succeeded: https://github.com/civiform/civiform-staging-deploy/actions/runs/3047271597/jobs/4911075726

Next up, I'll work on making the changes to the Seattle probers.

**Comment by clouser2** [comment]
Seattle staging had a successful prober run with the changes in place! https://github.com/seattle-civiform/civiform-deploy/actions/runs/3055080114/jobs/4927733573

I'll let this run again tomorrow and confirm success. Then we can clean up the fallback code in our test infrastructure image :D.

**Comment by clouser2** [comment]
Seattle staging probers have passed twice since the environment variable changes were made. AWS staging probers still seem to be failing sporaidcally in tests which exercise file upload. I'm suspicious that something about the test timing has changed recently which has made this worse. Fortunately, #3389 has a fix for this type of race condition where tests were clicking the "Upload" button and not properly waiting for the redirect to finalize.
